src/refda/exec.c

```diff
@@ -375,91 +376,260 @@ void *refda_exec_worker(void *arg)
+        if (!at_end)
+        {
+            refda_exec_exp_execset(agent, &item);
+        }
+        refda_msgdata_deinit(&item);
+        if (at_end && refda_timeline_empty_p(agent->exec_timeline))
         {
-            CACE_LOG_DEBUG("waiting for exec event");
-            sem_wait(&(agent->execs_sem));
+            CACE_LOG_INFO("Got undefined exec, stopping");
+
+            // flush the input queue but keep the daemon running
+            refda_msgdata_t undef;
+            refda_msgdata_init(&undef);
+            refda_msgdata_queue_push_move(agent->rptgs, &undef);
+            sem_post(&(agent->rptgs_sem));
+
+            return;
         }
+    }
+
+    // execute any waiting sequences
+    refda_exec_waiting(agent);
+}
+
+static int refda_exec_schedule_tbr(refda_agent_t *agent, refda_amm_tbr_desc_t *tbr, bool starting);
+
+/** Execute a time based rule's action that has already been verified
+ * Based on code from refda_exec_exp_execset
+ */
+static int refda_exec_tbr_action(refda_agent_t *agent, refda_exec_seq_t *seq, const refda_amm_tbr_desc_t *tbr)
+{
+    refda_runctx_ptr_t ctxptr;
+    refda_runctx_ptr_init_new(ctxptr);
+
+    refda_runctx_t *runctx = refda_runctx_ptr_ref(ctxptr);
+
+    if (refda_runctx_from(runctx, agent, NULL))
+    {
+        return 2;
+    }
+
+    refda_runctx_ptr_set(seq->runctx, ctxptr);
+    int res = refda_exec_exp_mac(runctx, seq, &(tbr->action));
+
+    refda_runctx_ptr_clear(ctxptr); // Clean up extra reference created by ptr_ref
+    return res;
+}
+
+/** Begin a single execution of a time based rule
+ */
+static void refda_exec_tbr(refda_agent_t *agent, refda_amm_tbr_desc_t *tbr)
+{
+    CHKERR1(agent);
+    CHKERR1(tbr);
+
+    if (!tbr->enabled)
+    {
+        CACE_LOG_INFO("TBR is not enabled");
+        return;
+    }
+
+    if (refda_amm_tbr_desc_reached_max_exec_count(tbr))
+    {
+        CACE_LOG_INFO("TBR reached maximum execution count");
+        return;
+    }
+
+    if (tbr->action.is_ref || tbr->action.as_lit.ari_type != CACE_ARI_TYPE_AC)
+    {
+        CACE_LOG_ERR("Invalid TBR action, unable to continue");
+        return;
+    }
 
-        // execs queue may still be empty if deferred callbacks were run
-        if (refda_msgdata_queue_pop(&item, agent->execs))
+    refda_exec_seq_t *seq = refda_exec_seq_list_push_back_new(agent->exec_state);
+    seq->pid              = agent->exec_next_pid++;
+
+    int exec_result = refda_exec_tbr_action(agent, seq, tbr);
+
+    if (!exec_result)
+    {
+        tbr->exec_count++;
+
+        // Schedule next execution of the rule
+        //
+        // FUTURE: Consider setup after the TBR's action is fully executed, rather than here.
+        //  Likely is fine either way but if action takes awhile and the period is short
+        //  it could cause problems with overlapping execution. On the other hand if that
+        //  happens the rule is almost certainly too aggressive.
+        refda_exec_schedule_tbr(agent, tbr, false);
```

Comment:
The scheduling should actually occur independently of the execution time, so that if the config is to execute every 2.0 seconds each subsequent start-to-start time should be exactly 2.0 seconds on the timeline.

There is agent execution telemetry available to know if there becomes a backlog of pending sequences.

Reply:
Cleaned up the comment - thanks for clarifying!